packages/kinvey-datastore/src/repositories/offline-repositories/inmemory-offline-repository.js

```diff
@@ -0,0 +1,203 @@
+import keyBy from 'lodash/keyBy';
+import { NotFoundError } from 'kinvey-errors';
+
+import { OfflineRepository } from '../offline-repository';
+import {
+  applyQueryToDataset,
+  collectionsMaster as masterCollectionName
+} from '../utils';
+import { ensureArray } from '../../utils';
+
+// Imported for typings
+// import { KeyValuePersister } from '../../persisters';
+
+export class InmemoryOfflineRepository extends OfflineRepository {
+  /** @type {KeyValuePersister} */
+  _persister;
+  /** @type {PromiseQueue} */
+  _queue;
+
+  constructor(persister, promiseQueue) {
+    super();
+    this._persister = persister;
+    this._queue = promiseQueue;
+  }
+
+  // ----- private methods
+
+  // TODO: integrate with db/collection concept in parent(s)?
+  _formCollectionKey(collection) {
+    const appKey = this._getAppKey();
+    return `${appKey}${collection}`;
+  }
+
+  _readAll(collection) {
+    const key = this._formCollectionKey(collection);
+    return this._persister.readEntities(key)
+      .then(entities => entities || []);
+  }
+
+  // TODO: Keep them by id
+  _saveAll(collection, entities) {
+    const key = this._formCollectionKey(collection);
+    return this._persister.persistEntities(key, entities);
+  }
+
+  _deleteAll(collection) {
+    const key = this._formCollectionKey(collection);
+    return this._persister.deleteEntities(key);
+  }
+
+  _enqueueCrudOperation(collection, operation) {
+    const key = this._formCollectionKey(collection);
+    return this._queue.enqueue(key, operation);
+  }
+
+  _getAllCollections() {
+    return this._readAll(masterCollectionName);
+  }
+
+  _countAndDelete(collection) {
+    return this._readAll(collection)
+      .then((entities) => {
+        return this._deleteAll(collection)
+          .then(() => entities.length);
+      });
+  }
+
+  _clearAllCollections() {
+    return this._getAllCollections()
+      .then((collections) => {
+        const promises = collections.map(c => this._deleteAll(c));
+        return Promise.all(promises);
+      })
+      .then(() => this._deleteAll(masterCollectionName));
+  }
+
+  // ----- protected methods
+  _ensureCollection(collection) {
+    return this._readAll(masterCollectionName)
+      .then((collectionsForAppKey) => {
+        const exists = collectionsForAppKey.indexOf(collection) > -1;
+        if (!exists) {
+          collectionsForAppKey.push(collection);
+          return this._saveAll(masterCollectionName, collectionsForAppKey);
+        }
+        return Promise.resolve();
+      });
+  }
+
+  // ----- public methods
+
+  // TODO: add checks for existing collection, try/catch, etc - here or in processor
+
+  create(collection, entitiesToSave) {
+    return this._enqueueCrudOperation(collection, () => {
+      return this._ensureCollection(collection)
+        .then(() => this._readAll(collection))
```

Comment:
Seems highly inefficient. Good place to target for subsequent performance improvement.

Reply:
Could you elaborate?

Reply:
The fact that we call `_readAll` at multiple places in this class seems inefficient. I know it's not in scope for the redesign, but it points to a performance issue that'll need to be solved eventually.

Reply:
This being an "inmemory" repository, there is no way around that. Any performance-intensive work should not be using this type of repository at all. But it is there for backward compatibility. The TTL we discussed in another comment was precisely for this. To avoid multiple reads, the read items are being cached inmemory. This is done in the inmemory repo's persister.